File: javascript/menstrual app/MenstrualApp.js
function calculateCycleDates(startingDateStr, cycleLength) {
	const startDate = new Date(startingDateStr);
	if (isNaN(startDate)) {
	throw new Error("Invalid date format. Use YYYY-MM-DD format");
	}

	const ovulationDay = new Date(startDate);
	ovulationDay.setDate(ovulationDay.getDate() + (cycleLength - 14));
	const fertileStart = new Date(ovulationDay);
	fertileStart.setDate(fertileStart.getDate() - 4);
	
	const fertileEnd = new Date(ovulationDay);
	fertileEnd.setDate(fertileEnd.getDate() + 1);

	const nextPeriod = new Date(startDate);
	nextPeriod.setDate(nextPeriod.getDate() + cycleLength);
	
	return {
	startDate,
	ovulationDay,
	fertileWindow: [fertileStart, fertileEnd],
	nextPeriod
	};
}

module.exports = {
	calculateCycleDates
	};

File: python/menstrual app/menstrual_app.py
from datetime import datetime, timedelta
def cycle_dates_calculation(starting_date, cycle_length):
	start_date = datetime.strptime(starting_date, "%Y-%m-%d").date()
	ovulation_day = start_date + timedelta(days=cycle_length - 14)
	fertile_window_open = ovulation_day - timedelta(days=4)
	fertile_window_close = ovulation_day + timedelta(days=1)
	next_period = start_date + timedelta(days=cycle_length)
	return{
		"start_date": start_date,
		"ovulation_day": ovulation_day,
		"fertile_window": (fertile_window_open, fertile_window_close),
		"next_period": next_period
	}